modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/metadata/IgniteMdRowCount.java

```diff
@@ -59,62 +67,125 @@ public class IgniteMdRowCount extends RelMdRowCount {
         return rel.estimateRowCount(mq);
     }
 
-    /** */
-    @Nullable public static Double joinRowCount(RelMetadataQuery mq, Join rel) {
-        if (!rel.getJoinType().projectsRight()) {
-            // Create a RexNode representing the selectivity of the
-            // semijoin filter and pass it to getSelectivity
-            RexNode semiJoinSelectivity =
-                RelMdUtil.makeSemiJoinSelectivityRexNode(mq, rel);
-
-            return multiply(mq.getSelectivity(rel.getLeft(), semiJoinSelectivity),
-                mq.getRowCount(rel.getLeft()));
-        }
+    /** Estimates rows number of a join product. If can't, falls back to Calcite's default implementation. */
+    public static @Nullable Double joinRowCount(RelMetadataQuery mq, Join join) {
```

Comment:
Let's try to safe variable names close to original code to reduce collisions later
Let's also get javadoc from original code.
What about IgniteCorrelatedNestedLoopJoin.estimateRowCount? It still divided by selectivity, do we really need it?